portal-impl/src/com/liferay/portlet/internal/PortletResponseImpl.java

```diff
@@ -446,17 +449,63 @@ public PortletRequestImpl getPortletRequest() {
 
 	@Override
 	public String getProperty(String key) {
-		throw new UnsupportedOperationException();
+		Object value = _headers.get(key);
+
+		if (value == null) {
+			return null;
+		}
+
+		if (value instanceof Object[]) {
+			Object[] values = (Object[])value;
+
+			value = ArrayUtil.getValue(values, 0);
+		}
+
+		if (value == null) {
+			return null;
+		}
+
+		return value.toString();
 	}
 
 	@Override
 	public Collection<String> getPropertyNames() {
-		throw new UnsupportedOperationException();
+		Set<String> keySet = _headers.keySet();
+
+		if (keySet == null) {
+			return Collections.emptySet();
+		}
+
+		return new LinkedHashSet<>(keySet);
 	}
 
 	@Override
 	public Collection<String> getPropertyValues(String key) {
-		throw new UnsupportedOperationException();
+		Object value = _headers.get(key);
+
+		if (value == null) {
+			return Collections.emptySet();
+		}
+
+		List<String> values = new ArrayList<>();
+
+		if (value instanceof Object[]) {
+			Object[] valueArray = (Object[])value;
```

Comment:
@ngriffin7a Isn't this breaking the api contract? The api says 

> This method considers only response properties set or added via setProperty(java.lang.String, java.lang.String) or addProperty(java.lang.String, java.lang.String)

All other headers will end up being an Object[] and passing the if check, then getting converted to string for returning.

I think you meant to do "instanceof String[]", right?

Reply:
@shuyangzhou Good catch. Would you like me to make the change and send a new PR?

Reply:
Yes, please